File: sdk/golang/evolution_outbox.go
package prismer

import (
	"context"
	"fmt"
	"math/rand"
	"sync"
	"time"
)

// EvolutionOutboxOp represents a queued evolution operation.
type EvolutionOutboxOp struct {
	ID             string
	OpType         string // "record" or "report"
	Payload        interface{}
	Status         string // "pending", "inflight", "confirmed", "failed"
	CreatedAt      time.Time
	Retries        int
	MaxRetries     int
	IdempotencyKey string
}

// EvolutionOutbox provides fire-and-forget outcome recording with local WAL.
//
// Queues Record() calls locally and flushes them to the server asynchronously.
// Prevents data loss on network failures and eliminates write latency from the
// agent's hot path.
//
// Usage:
//
//	outbox := prismer.NewEvolutionOutbox(requestFn, nil)
//	outbox.Start()
//	outbox.Enqueue("record", payload)
//	// ... later ...
//	outbox.Stop() // flushes remaining
type EvolutionOutbox struct {
	mu            sync.Mutex
	queue         []*EvolutionOutboxOp
	requestFn     func(ctx context.Context, method, path string, body interface{}) error
	flushInterval time.Duration
	maxRetries    int
	batchSize     int
	stopCh        chan struct{}
	running       bool
}

// EvolutionOutboxConfig configures the outbox behavior.
type EvolutionOutboxConfig struct {
	FlushInterval time.Duration // default 1s
	MaxRetries    int           // default 5
	BatchSize     int           // default 10
}

// NewEvolutionOutbox creates a new outbox with the given request function and config.
// The requestFn is called for each flush operation with (ctx, "POST", endpoint, payload).
func NewEvolutionOutbox(
	requestFn func(ctx context.Context, method, path string, body interface{}) error,
	cfg *EvolutionOutboxConfig,
) *EvolutionOutbox {
	if cfg == nil {
		cfg = &EvolutionOutboxConfig{}
	}
	if cfg.FlushInterval == 0 {
		cfg.FlushInterval = time.Second
	}
	if cfg.MaxRetries == 0 {
		cfg.MaxRetries = 5
	}
	if cfg.BatchSize == 0 {
		cfg.BatchSize = 10
	}
	return &EvolutionOutbox{
		requestFn:     requestFn,
		flushInterval: cfg.FlushInterval,
		maxRetries:    cfg.MaxRetries,
		batchSize:     cfg.BatchSize,
		stopCh:        make(chan struct{}),
	}
}

// Enqueue adds an operation to the outbox. Returns the operation ID.
func (o *EvolutionOutbox) Enqueue(opType string, payload interface{}) string {
	op := &EvolutionOutboxOp{
		ID:             fmt.Sprintf("evo_%d_%s", time.Now().UnixMilli(), randHex(8)),
		OpType:         opType,
		Payload:        payload,
		Status:         "pending",
		CreatedAt:      time.Now(),
		MaxRetries:     o.maxRetries,
		IdempotencyKey: fmt.Sprintf("%s:%d", opType, time.Now().UnixMilli()),
	}
	o.mu.Lock()
	o.queue = append(o.queue, op)
	o.mu.Unlock()
	return op.ID
}

// PendingCount returns the number of pending operations.
func (o *EvolutionOutbox) PendingCount() int {
	o.mu.Lock()
	defer o.mu.Unlock()
	count := 0
	for _, op := range o.queue {
		if op.Status == "pending" {
			count++
		}
	}
	return count
}

// Start begins the background flush goroutine.
func (o *EvolutionOutbox) Start() {
	o.mu.Lock()
	o.running = true
	o.mu.Unlock()
	go o.flushLoop()
}

// Stop stops the background flush goroutine and performs a final flush.
func (o *EvolutionOutbox) Stop() {
	o.mu.Lock()
	wasRunning := o.running
	o.running = false
	o.mu.Unlock()
	if wasRunning {
		close(o.stopCh)
	}
	o.Flush(context.Background())
}

// Flush sends pending operations to the server. Returns count of flushed ops.
func (o *EvolutionOutbox) Flush(ctx context.Context) int {
	o.mu.Lock()
	var pending []*EvolutionOutboxOp
	for _, op := range o.queue {
		if op.Status == "pending" && len(pending) < o.batchSize {
			pending = append(pending, op)
		}
	}
	o.mu.Unlock()

	flushed := 0
	for _, op := range pending {
		op.Status = "inflight"
		endpoint := "/api/im/evolution/record"
		if op.OpType == "report" {
			endpoint = "/api/im/evolution/report"
		}
		err := o.requestFn(ctx, "POST", endpoint, op.Payload)
		if err != nil {
			op.Retries++
			if op.Retries >= op.MaxRetries {
				op.Status = "failed"
			} else {
				op.Status = "pending"
			}
		} else {
			op.Status = "confirmed"
			flushed++
		}
	}

	// Remove confirmed/failed ops
	o.mu.Lock()
	var remaining []*EvolutionOutboxOp
	for _, op := range o.queue {
		if op.Status != "confirmed" && op.Status != "failed" {
			remaining = append(remaining, op)
		}
	}
	o.queue = remaining
	o.mu.Unlock()

	return flushed
}

func (o *EvolutionOutbox) flushLoop() {
	ticker := time.NewTicker(o.flushInterval)
	defer ticker.Stop()
	for {
		select {
		case <-o.stopCh:
			return
		case <-ticker.C:
			o.Flush(context.Background())
		}
	}
}

func randHex(n int) string {
	const chars = "0123456789abcdef"
	b := make([]byte, n)
	for i := range b {
		b[i] = chars[rand.Intn(len(chars))]
	}
	return string(b)
}

File: sdk/python/prismer/evolution_outbox.py
"""Evolution Outbox — fire-and-forget outcome recording with local WAL.

Queues record() calls locally and flushes them to the server asynchronously.
Prevents data loss on network failures and eliminates write latency from the agent's hot path.
"""

import threading
import time
import uuid
from collections import deque
from typing import Any, Callable, Deque, Dict, List, Optional


class EvolutionOutboxOp:
    """A single outbox operation."""

    __slots__ = ("id", "op_type", "payload", "status", "created_at", "retries", "max_retries", "idempotency_key")

    def __init__(self, op_type: str, payload: Dict[str, Any], max_retries: int = 5):
        self.id = f"evo_{int(time.time()*1000)}_{uuid.uuid4().hex[:8]}"
        self.op_type = op_type  # 'record' or 'report'
        self.payload = payload
        self.status = "pending"
        self.created_at = time.time()
        self.retries = 0
        self.max_retries = max_retries
        self.idempotency_key = f"{op_type}:{payload.get('gene_id', '')}:{int(time.time()*1000)}"


class EvolutionOutbox:
    """In-memory outbox with background flush thread.

    Usage:
        outbox = EvolutionOutbox(request_fn)
        outbox.start()
        outbox.enqueue('record', {'gene_id': '...', 'signals': [...], ...})
        # ... later ...
        outbox.stop()  # flushes remaining
    """

    def __init__(
        self,
        request_fn: Callable,
        flush_interval: float = 1.0,
        max_retries: int = 5,
        batch_size: int = 10,
    ):
        self._request_fn = request_fn
        self._flush_interval = flush_interval
        self._max_retries = max_retries
        self._batch_size = batch_size
        self._queue: Deque[EvolutionOutboxOp] = deque()
        self._lock = threading.Lock()
        self._timer: Optional[threading.Timer] = None
        self._running = False

    @property
    def pending_count(self) -> int:
        with self._lock:
            return sum(1 for op in self._queue if op.status == "pending")

    def enqueue(self, op_type: str, payload: Dict[str, Any]) -> str:
        """Add an operation to the outbox. Returns the operation ID."""
        op = EvolutionOutboxOp(op_type, payload, self._max_retries)
        with self._lock:
            self._queue.append(op)
        return op.id

    def start(self) -> None:
        """Start the background flush timer."""
        self._running = True
        self._schedule_flush()

    def stop(self) -> None:
        """Stop the background flush and do a final flush."""
        self._running = False
        if self._timer:
            self._timer.cancel()
            self._timer = None
        self.flush()

    def flush(self) -> int:
        """Flush pending operations. Returns count of successfully flushed ops."""
        with self._lock:
            pending = [op for op in self._queue if op.status == "pending"][:self._batch_size]

        flushed = 0
        for op in pending:
            op.status = "inflight"
            try:
                endpoint = "/api/im/evolution/record" if op.op_type == "record" else "/api/im/evolution/report"
                self._request_fn("POST", endpoint, json=op.payload)
                op.status = "confirmed"
                flushed += 1
            except Exception:
                op.retries += 1
                if op.retries >= op.max_retries:
                    op.status = "failed"
                else:
                    op.status = "pending"

        # Remove confirmed/failed ops
        with self._lock:
            self._queue = deque(op for op in self._queue if op.status not in ("confirmed", "failed"))

        return flushed

    def _schedule_flush(self) -> None:
        if not self._running:
            return
        self._timer = threading.Timer(self._flush_interval, self._flush_and_reschedule)
        self._timer.daemon = True
        self._timer.start()

    def _flush_and_reschedule(self) -> None:
        try:
            self.flush()
        except Exception:
            pass
        self._schedule_flush()
